0015-estimator-interface.md

```diff
@@ -0,0 +1,414 @@
+
+```python
+def coerce(argument: TypeLike) -> Type
+```
+
+that is invoked inside of the `run()` method.
+
+For example,
+
+```python
+BindingsArrayLike = Union[BindingsArray, ArrayLike, Mapping[Parameter, float]]
+```
+
+and
+
+```python
+@staticmethod
+def coerce(bindings_array: BindingsArrayLike) -> BindingsArray:
+    if isinstance(bindings_array, (list, ArrayLike)):
+        bindings_array = BindingsArray(bindings_array)
+    elif isinstance(bindings_array, Mapping):
+        bindings_array = BindingsArray(kwargs=bindings_array)
+
+    return bindings_array
+```
+
+In particular, we propose this kind of Coercion for the types:
+* `EstimatorTask`
+* `BindingsArray`
+* `ObservablesArray`
+
+### TaskResults <a name="TaskResults"></a>
+
+The results from each `EstimatorTask` will be array valued, and each `EstimatorTask` in the same job may have a different shape.
+Consider an `EstimatorTask` with shape `<20, 30>`, where the shape has come from the broadcasting rules discussed elsewhere. 
+This will result in a 20×30 array of real estimates.
+Moreover, we will want to return an array of standard deviations of the same shape.
+This would result in a bundle of broadcastable arrays:
+
+```python
+TaskResult({“evs”: <20, 30>, “stds”: <20, 30>}, metadata)
+```
+
+The reason we are proposing a generic container for the return type instead of, e.g., an `Estimator`-specific container, is because it
+
+1. Provides unified experience across primitives for users. 
+2. code-reuse
+3. Provides a certain certain amount of flexibility for what can be returned without modifying the container object. Here are some examples:
+  1. Suppose that we want to give users the option of additionally returning the covariances between estimates that arise because of the circuit multiplexing, then we could update with the field `{"cov": <20,30,20,30>}`.
+  2. Suppose we want to return some indication of which estimates came from the same physical circuit.
+
+## Migration Path <a name="migration-path"></a>
+
+We need to remain backwards compatible with the existing interface to adhere to the [Qiskit Deprecation Policy](https://qiskit.org/documentation/deprecation_policy.html).
+
+In summary, we propose the following strategy: explicitly version the `Estimator` interface (i.e. add a `BaseEstimatorV2` as described in this proposal) and document both the differences and the end user migration path for consumers of estimators. The user can determine which version of the estimator to use, and existing implementations can be left intact.
+
+Similar to `Backend`, we will have a `BaseEstimator`, `BaseEstimatorV1` equal to the existing `BaseEstimator`, and `BaseEstimatorV2`, as described in this proposal.
+We will maintain backward compatibility by defaulting the import of `Estimator` to `EstimatorV1` (i.e. existing) for the duration of the deprecation period.
```

Comment:
Which `Estimator` are you talking about here? I assume `qiskit.primitives.Estimator`, but I know of at least three classes named `Estimator` that implement `BaseEstimator` in different repositories.

Regardless of which one is being referred to, one other option here that might make it easier for users, is basically what `qiskit-ibm-provider` does for its `BackendV2` implementation. The backends from `qiskit-ibm-provider` implement the `BackendV2` interface but also provide a compatible interface with most of `BackendV1` (the notable exception being `backend.name` which is an attribute on V2 and a method on V1). So users have the freedom to use both interfaces for an existing implementation. It also makes future deprecation a little easier because it's easier to warn on legacy access patterns and inputs. But it's also totally manageable with parallel implementations, just using a future warning when `Estimator` is instantiated and you can use a `__future__` import to change `Estimator` over to `EstimatorV2`.